pkg/agent/openflow/feature.go

```diff
@@ -0,0 +1,254 @@
+
+// feature is the interface to program a major function in Antrea data path. The following structures implement this interface:
+// - featurePodConnectivity, implementation of connectivity for Pods, activated by default.
+// - featureNetworkPolicy, implementation of K8s NetworkPolicy and Antrea NetworkPolicy, activated by default.
+// - featureService, implementation of K8s Service, activated by default.
+// - featureEgress, implementation of Egress, activation is determined by feature gate Egress.
+// - featureMulticast, implementation of multicast, activation is determined by feature gate Multicast,
+// - featureTraceflow, implementation of Traceflow.
+type feature interface {
+	// getFeatureName returns the name of the feature.
+	getFeatureName() string
+	// getRequiredTables returns a slice of required tables of the feature.
+	getRequiredTables() []*Table
+	// initFlows returns the initial flows of the feature.
+	initFlows() []binding.Flow
+	// replayFlows returns the fixed and cached flows that need to be replayed after OVS is reconnected.
+	replayFlows() []binding.Flow
+}
+
+// Table in FlexiblePipeline is the basic unit to build OVS pipelines. A Table can be referenced by one or more features,
+// but its member struct ofTable will be initialized and realized on OVS only when it is referenced by any activated features.
+type Table struct {
+	name     string
+	stage    binding.StageID // All stages are defined in pkg/ovs/openflow/interfaces.go
+	pipeline binding.PipelineID
+	ofTable  binding.Table
+}
+
+// tableOrderCache is used to save the order of all defined tables located in file pkg/agent/openflow/pipeline.go. The tables
+// have the same pipeline ID are saved in a slice by order of definition. When building a pipeline, the required table list
+// is aggregated from all activated features, and the order of the required tables is decided by the saved order.
+var tableOrderCache map[binding.PipelineID][]*Table
+
+// newTable is used to declare a Table. A table should belong to a stage and pipeline defined in file pkg/ovs/openflow/interfaces.go.
+// Stage is used to group tables which implement similar functions in a pipeline. Stages include:
+//   - ClassifierStage, implementation of classifying packet "category" (tunnel, local gateway or local Pod, etc).
+//   - ValidationStage, implementation of validating packets.
+//   - ConntrackStateStage, implementation of transforming committed packets in CT zones.
+//   - PreRoutingStage, similar to PREROUTING chain of nat table in iptables. DNAT of Service connection is performed in
+//     this stage.
+//   - EgressSecurityStage, implementation of egress rule for K8s NetworkPolicy and Antrea NetworkPolicy.
+//   - RoutingStage, implementation of forwarding packets in L3.
+//   - PostRoutingStage, similar to POSTROUTING chain of nat table in iptables. SNAT of Service connection is performed
+//     in this stage.
+//   - SwitchingStage, implementation of "dmac" table.
```

Comment:
What does "dmac" table mean? I think we do not have a table called "dmac". Do you mean "L2 forwarding of packets"?

Reply:
Yes, L2 forwarding of packets